### DISCUSSION TITLE
Editing Records within a View

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#653] [Maintainer discussions]
The design for issue #456  is ready for review. Please follow the [Review Guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines) and make sure to review before the set deadline expires.

### Document
https://wiki.mathesar.org/en/design/specs/eding-view-records

### Reviewers
@kgodey @pavish @mathemancer  Your review is required.
Feedback from the community is also welcome. Contributors can find more context about Mathesar's design strategy [here](https://wiki.mathesar.org/design).

### Deadline
The deadline for submitting feedback is set 2 days from now, on Tuesday 14.

**Comment by kgodey** [comment]
@ghislaineguerin I have a bunch of questions/comments:

(1) What is the logic for when we show the form and when we allow the user to edit-in-place while adding a record? Can the user access the form if they prefer to use that?

(2) I don't see any design for removing a record in the prototype. Also, while designing this, we need to address the potential confusion a user might have about deleting records from views vs. tables. They may think they are only deleting a row from the view but we are deleting all records referencing the row from the underlying table. We need some form of confirmation that says "this will delete record X from table A and record Y from table B, are you sure?"

This is not as much of an issue for adding records because they can always "undo" adding records by deleting them, but they cannot do the opposite.

Also, just in case this is not clear, they will be deleting an entire row, they cannot delete just one record from the row if each row consists of multiple records.

(3) The non-updatable view looks good for cases where we can find the underlying tables. In some cases, we may not be able to identify which tables a column comes from automatically, so those won't be able to be edited at all. We need a design for this too.

(4) Generally, I think in views, we need to think about rows vs. records. A row may be composed of multiple (partial) records or just one record. This goes for adding and deleting (not so much for editing existing values).

(5) I still don't see an indicator for unsaved changes. I might be missing it if it's there.

(6) The column drag and drop looks good, but we also need an indicator for changes not being saved. This might be the same feedback as (5).

(7) I'm not sure what the "Global Updates" section is referring to and I'm not sure how to review it.

**Reply by ghislaineguerin** [reply]
Thanks @kgodey 
>(1) What is the logic for when we show the form and allow the user to edit-in-place while adding a record? Can the user access the form if they prefer to use that?

Users can always access the form by clicking on the 'Edit' option in the view toolbar whenever a record field is selected. 

>(2) I don't see any design for removing a record in the prototype. Also, while designing this, we need to address the potential confusion a user might have about deleting records from views vs. tables. They may think they are only deleting a row from the view, but we delete all records referencing the row from the underlying table. We need some form of confirmation that says, "this will delete record X from table A and record Y from table B, are you sure?"

Got it. I've added this now. I'm wondering if there's a chance for other related records to be cascade deleted:
"A foreign key with cascade delete means that if a record in the parent table is deleted, then the corresponding records in the child table will automatically be deleted."

>This is not an issue for adding records because they can always "undo" adding records by deleting them, but they cannot do the opposite. Also, just in case this is not clear, they will be deleting an entire row. They cannot delete just one record from the row if each row consists of multiple records.

It is clear that the row is being deleted. Maybe we need to represent the row selection visually. I will experiment with this.

>(3) The non-updatable view looks good for cases where we can find the underlying tables. In some cases, we may not be able to identify which tables a column comes from automatically, so those won't be edited at all. We need a design for this too.

Makes sense. I have included this as part of the design. I think we'd have to remove the color references from the columns as well. Also, in the menu, we would only be able to show the column that is being referenced but can't link to the actual table.

>(4) Generally, I think in views, we need to think about rows vs. records. A row may be composed of multiple (partial) records or just one record. This goes for adding and deleting (not so much for editing existing values).

I assume that while editing, the user might clear a record, meaning the user hits delete and the field becomes empty, in which case they might replace the values. For example, given that we have a view like:
https://www.figma.com/proto/Uaf1ntcldzK2U41Jhw6vS2/Mathesar-MVP?page-id=4928%3A47261&node-id=5072%3A57556&viewport=415%2C48%2C0.27&scaling=contain&starting-point-node-id=5118%3A66661&show-proto-sidebar=1
Where the playlist table contains a list of track references (foreign key) with a play order field. And the 'trackName' column from the track table, is being referenced from that foreign key. 
If a user wanted the second track 'Around the World' to play first, they'd have to edit the playlist record and assign a different order. However, my concern is that if they edit the actual name on the track record that is currently in position 1, in this case 'Turn Back Time' and they rename it, they'd just be changing the name of that track, which might cause other problems. We need to think about editing without the context of the form and if it could lead to confusion. 

>(5) I still don't see an indicator for unsaved changes. I might be missing it if it's there.

I've added the unsaved dot next to the view title, but I will make this more prominent.

>(6) The column drag and drop looks good, but we also need an indicator for changes not being saved. This might be the same feedback as (5).

I will update it so that it is more prominent.

>(7) I'm not sure what the "Global Updates" section refers to, and I'm not sure how to review it.

I'm adding more details to the document. These are the prototype components that are part of the same milestone but are now consolidated into this prototype.

**Reply by kgodey** [reply]
> Got it. I've added this now. I'm wondering if there's a chance for other related records to be cascade deleted:
"A foreign key with cascade delete means that if a record in the parent table is deleted, then the corresponding records in the child table will automatically be deleted."

The deletion confirmation looks good. I would add something like "if you want to remove it from this view, try adjusting the filters" or something like that.

Currently, we don't allow any cascade deletions, we show an error saying the record cannot be deleted if something else depends on it. So that shouldn't be a problem.

> Maybe we need to represent the row selection visually. I will experiment with this.

Sounds good.

> Makes sense. I have included this as part of the design. I think we'd have to remove the color references from the columns as well. Also, in the menu, we would only be able to show the column that is being referenced but can't link to the actual table.

Looks good, yes, we should delete the color. I think we should also remove the "referenced from" section entirely since it just repeats the column name and provides no additional information.

> If a user wanted the second track 'Around the World' to play first, they'd have to edit the playlist record and assign a different order. However, my concern is that if they edit the actual name on the track record that is currently in position 1, in this case 'Turn Back Time' and they rename it, they'd just be changing the name of that track, which might cause other problems. We need to think about editing without the context of the form and if it could lead to confusion.

I agree it could lead to confusion. Do you think we should disable editing-in-place and only allow editing via the form?

My instinct is to allow editing in place and test it with real users to see how they react, the convenience might be worth it.

> I'm adding more details to the document. These are the prototype components that are part of the same milestone but are now consolidated into this prototype.

Okay, so I should review the listed scenarios in the Create View prototype. It all looks good to me. Maybe you could update the other discussions and say they're being closed in favor of this one?

**Comment by mathemancer** [comment]
Overall, I think this makes sense.  I like the record editing sidebar.  I have a couple concerns:
- If the user is editing directly in the cell, I don't think it will be obvious enough to some more novice users that they're actually editing a different object (the table containing the record referenced by the cell).  I think they should _always_ be made aware of precisely which record is being edited (for example, the view might have multiple cells in a column with the same value, but from different records in some underlying table.  They could be quite confused about which they're editing if they don't see the actual record in question).
- I don't have a clear picture of how the non-editable view will be defined, and how the referenced table would be determined.  The point of editable vs. non-editable from a DB point of view is that a view is non-editable precisely if you _can't_ logically define the underlying table or record.  For example, if the view contains aggregations of some kind.  If a column is composed of the mean of some underlying column, grouped by some other key, there isn't a definable place to edit to change cells in that column.  There may not even be a definable table to navigate to (what if the cell is an aggregation of data from more than one table?).  With that in mind, I don't see the point of different flows between "editable" and "non-editable".  The "editable" flow (with the record editing sidebar) already handles all scenarios I can think of where the source for a given cell (or row) in a view can actually be defined.  I.e., any scenarios where conflicts could arise due to constraints in the underlying table already seem to be handled.

**Comment by pavish** [comment]
I was looking through the combined view flow in the above UX prototype, and it looks pretty good. I have a few concerns:

1. I do think for views, it would be essential to show the form of the linked tables. However, I think we can experiment not using a sidebar and showing it as a dropdown (using ample horizonal space along with vertical) below each cell to provide the feeling of inplace editing.

2. Adding a new record in a view
  A view row can consist of partial records from multiple tables/other views. When adding a new row in a view, we can add a new one on the table or use an existing one. There currently seems to be no flow for using an existing record from linked tables.

3. Editing cells
  What does non-editable view entail? If we are showing the form, it covers the details needed from the linked tables. In certain cases like primary keys and aggregated cells, we can indicate in the cell that they cannot be edited.

**Comment by ghislaineguerin** [comment]
@kgodey regarding the non-editable views. I assume since we can't reference the tables nor add new columns or records, users will have limited functionality, such as filter, sort, and group. However, I'm not sure if we allow them to save that view with the filters as that would effectively 'edit' the view, wouldn't it?

**Reply by kgodey** [reply]
@ghislaineguerin If a view is not editable, I'm not sure that we can even support filter, sort, and group. We need to do some backend implementation first before we have the answer to that.

However, if we do support filter, sort, and group, then we can also save it (since both operations involve updating the database query used to generate the view).

**Reply by ghislaineguerin** [reply]
@kgodey I will assume that the view is ready only with no other functionality then.

**Reply by mathemancer** [reply]
@ghislaineguerin @kgodey I think there's some confusion about "non-editable".  Unless I'm completely confused about the definition we're using here, the things that would be restricted for a non-editable view are inserting and updating records (and dropping, but you could simulate this with filters).  You should still be able to do things like sort, group, and filtering.  This is because you're not inserting or updating records via any of those operations, but rather transforming your view of records that already exist in the database.

The one thing mentioned in this thread that I'd need to think more about would be adding columns to a view.  This would be possible as long as the column is from a table that's already joined into the view in some way, otherwise you'd need to define some join condition to add columns.

**Comment by ghislaineguerin** [comment]
@kgodey I want to confirm the behavior for editing records based on our discussion yesterday:
#### Editing or Adding Records to a View
A view with columns from multiple tables will display the table references in the new record row. Each cell will specify the table to which a record is being added.

![image](https://user-images.githubusercontent.com/845767/136082472-e678d319-e134-4200-87c9-958829c9be97.png)

#### The edit form
Editing any of the cells will prompt the record form and allow the user to edit all of the record fields for that table.

![image](https://user-images.githubusercontent.com/845767/136082742-d4e6ae10-07e1-43c2-aae8-ab152783bd02.png)

#### Question
In this scenario, the foreign key for 'releaseName' exists in the 'artist' table as 'releaseId'
The user can modify the 'releaseId' field from here which would change the 'releaseName' associated. However, the records from 'releaseName' cannot be swapped since there's not foreign key field on that table. 

Should we allow them to swap a record? Meaning that we'd be modifying the 'artist' table?

![image](https://user-images.githubusercontent.com/845767/136083868-54cf62f0-78f5-415c-9b89-cab7beed2950.png)

Also, to confirm, do we want to have cancel and update buttons for dropdown actions or save on close?

**Reply by kgodey** [reply]
@ghislaineguerin The behavior you described seems accurate to me.

> In this scenario, the foreign key for 'releaseName' exists in the 'artist' table as 'releaseId'
> The user can modify the 'releaseId' field from here which would change the 'releaseName' associated. However, the records from 'releaseName' cannot be swapped since there's not foreign key field on that table.
> 
> Should we allow them to swap a record? Meaning that we'd be modifying the 'artist' table?

If they modify the release ID, then the release name in the View will automatically be updated. And that's fine, it's the desired behavior.

I'm not sure what you mean by "swapping" a record.

**Comment by kgodey** [comment]
Based on the new action plan from the [2021-10-15 meeting](https://wiki.mathesar.org/meeting-notes/october-2021), we're going to update the requirements for issues in the milestone. I'm marking this discussion done, we'll start new discussions for further updates.
